Add new classes/functions to select and modify text as text (not structure)

arnt opened the pull request:
The new cookbook entries would be along the lines of.

To make all instances of the word jsoup into a link:
```
    html.find("jsoup").splitByBlockElements().wrap("<a href=http://www.jsoup.org>");
```

To make sure a particular string is a block element and has a particular class:
```
    html.find(foo.bar()).wrap("<div class=particular>);
```

I was a little uncertain how to name the getters. Much of jsoup uses foo(), but e.g. TextNode.getWholeText() uses getFoo(), like the more modern parts of the java standard library. I picked e.g. Point.getOffset() now, but I'll be happy to change to offset() if you prefer that.

cketti commented:
This is quite a substantial increase in API surface. Did you write down your motivation for adding such functionality somewhere?

I have the feeling (but no data to back this up) that most people use jsoup in a way that doesn't require such functionality. From that point of view it's only added complexity and a maintenance burden.
But I can certainly see that some people may want this. Maybe this could live in a companion library?

arnt commented:
Yes.

I have two needs (plus variations), and I added a little more for API roundness (see below).

1. I deal with HTML that contains undeclared boilerplate/structure. To handle with that I locate the boilerplate and wrap new div+CSS around it.

2. I need to elide because of output space restrictions. To achieve that I work with the text, identify and rank elision candidates, then mark those candidates with CSS and send javascript along to elide just as much as needed.

Then I wrote class descriptions, and added enough functions to make  them sound sensible. For example, it sounded strange to have Region.wrap(), Regions.wrap() and Regions.remove() but no Region.remove(), so I added the fourth merely to make the documentation sound smooth.

The Point class and Region.getStart()/getEnd() are public for the same reason, those could be package-private.

Regions.first() exists because e.g. Elements.first() does, it could be dropped and the callers could use get(0) instead. I don't personally see the reason to add Region.equals() and Region.hashCode(), but they're good form so I added them.

I agree that few people are using jsoup do do these things: It's a pain. But when something is a pain, even people who need the functionality don't write the code. "We need this, but it's a lot of work... let's put the ticket on the backlog".

jhy commented:
Thanks. I'm not really sure how developers would use this, and what the demand:complexity ratio is (to @chikei's point). Could you maybe show a couple examples of the input/output and code with this or with existing jsoup?

I'm wondering if we can make a few potentially simpler changes to improve text handling (maybe a wrapRange method) that hits maybe 80% of your use case with 20% of the code?

arnt commented:
Everything I've used this for falls into a few basic patterns. I start with text that's a mixture of something entered into a system and something supplied by the software. The text is not very disciplined and can be lacking in structure,. there can be nice structure around it. Overall the using code looks like some variation of this:

1. Get some HTML. Retrieve text(), sometimes of the Document, sometimes of a selected()ed Element.
2. Look at the text as text, identify patterns, boilerplate, etc. At the end of this you have a number of known strings. If you start by looking for EANs, now you have a set of strings that are valid EANs in that particular document.
3. For each of those strings, look for the corresponding Region in the source document and do the right thing. Sometimes the regions between identified strings are the interesting parts (for example the HTML between the end of an EAN and the end of the surrounding block element, but at most 500 characters).

Now, if you're looking for article number and description in a somewhat upgefucked site, and you know that the article number is an x-digit number in a table cell and the description is one or more paragraphs and perhaps a table and/or image following the number, then jsoup doesn't help to do much with that. If you also know that the article-number cell is wrapped in div class=articlenumber and the description is wrapped in div class=description, fine, but that's precisely what I didn't know. I knew things about the textual content, and (depending on use case) I needed to either transform the input and impose div structure, or to identify and extract parts of it (including images, tables, links and bold/italic blah).

It's possible that noone wants to use jsoup for that kind of thing because noone in their right mind works with messy and erratically formatted input. If so, fine, I've offered the patch, feel free to reject it. No hard feelings.

It's also possible that noone uses jsoup for that kind of thing because it's really hard to do without something like this patch. In that case you have to decide whether you want jsoup to be a suitable tool for such things or not.

I tried to think about a nice public example... perhaps scraping craigslist? The people who enter data there are undisciplined, so the user-supplied part of each page is messy, but there are common stock phrases, patterns, conventions.

I saw this recently: https://machinelearningblogs.com/2017/01/26/text-clustering-get-quick-insights-from-unstructured-data/ That sounds like what I've been doing, in a way... process (some of) the text as text, but then process it as HTML afterwards. So you want to add boldface to, or extract some valid HTML for, a passage that you identify using text clustering, machine learning, whatever? Then work on the text, make a Region, perhaps change the Region a little if either end is close to a paragraph boundary, then mark/clone it using Element functions.

Less API would be possible; Element.wrapRange(int, int), .cloneRange(int, int), .removeRange(int, int), List<IntPair> Element.splitRangeByBlockElement(int, int) should handle most of it. That's more or less how I started. Some things I needed can't be done using that. For example sometimes images at a region border should be in the wrapped/removed/cloned part and sometimes not. Sometimes I needed the block element(s) surrounding a given text: "Identify key phrase, usually but not always within one block element, clone that/those surrounding block elements". Sometimes I needed the identified string plus a few surrounding words and/or the rest of the block element, whichever is shorter. I threw away the int-based patch and wrote it again with Point instead of an int and several ways to set it, and a Region to do the wrapping/cloning/etc.

Again, feel free to reject. Some PRs are real requests, this is more of an offer.

jhy commented:
Thanks @arnt. I'll close this one per the discussion, but I do see the use case, and think it might be best as a separate companion library, so please do consider making that.

arnt commented:
Thanks for closing it.

That may sound like a strange thing to say, but IMO rejecting a pull request is okay while leaving old PRs for years is not. I'll buy you a beer if you're ever in Munich. Close some more and I'll buy you two.

FWIW I cannot do this as a separate library as it stands, it really needs [one new function (in Node)](https://github.com/jhy/jsoup/pull/833/files#diff-605d28890f72a0f43298f842d0a3414fR622) and would be horribly contorted without it. I could do it as a separate library if you were willing to merge just that function.
